File: php/src/CodingInterview/Application/Usecase/Portfolio/GetLatestPortfolioUsecase.php
<?php

declare(strict_types=1);

namespace Folio\CodingInterview\Application\Usecase\Portfolio;

use Folio\CodingInterview\Application\Repository\PortfolioRepository;
use Folio\CodingInterview\Domain\BigDecimal;
use Folio\CodingInterview\Domain\StockSymbol;

final class GetLatestPortfolioItemOutput
{
    public function __construct(
        public readonly StockSymbol $symbol,
        public readonly BigDecimal $rate,
    ) {}
}

final class GetLatestPortfolioUsecaseOutput
{
    /** @param GetLatestPortfolioItemOutput[] $items */
    public function __construct(public readonly array $items) {}
}

final class GetLatestPortfolioUsecase
{
    public function __construct(private readonly PortfolioRepository $portfolioRepository) {}

    public function run(): GetLatestPortfolioUsecaseOutput
    {
        $p = $this->portfolioRepository->get();
        $items = [];
        foreach ($p->items as $i) {
            $items[] = new GetLatestPortfolioItemOutput($i->symbol, $i->rate);
        }
        return new GetLatestPortfolioUsecaseOutput($items);
    }
}

File: scala/src/main/scala/folio/codinginterview/application/usecase/portfolio/GetLatestPortfolioUsecase.scala
package folio.codinginterview.application.usecase.portfolio

import folio.codinginterview.application.repository.PortfolioRepository
import folio.codinginterview.domain.StockSymbol
import folio.codinginterview.domain.Portfolio
import scala.concurrent.ExecutionContext
import scala.concurrent.Future

case object GetLatestPortfolioUsecaseInput

final case class GetLatestPortfolioItemOutput(symbol: StockSymbol, rate: BigDecimal)

final case class GetLatestPortfolioUsecaseOutput(items: Seq[GetLatestPortfolioItemOutput])

sealed trait GetLatestPortfolioUsecaseException extends RuntimeException

final class GetLatestPortfolioUsecase(
    portfolioRepository: PortfolioRepository
)(using ec: ExecutionContext) {
  def run(): Future[GetLatestPortfolioUsecaseOutput] = {
    portfolioRepository.get().map { p =>
      GetLatestPortfolioUsecaseOutput(
        p.items.map(i => GetLatestPortfolioItemOutput(i.symbol, i.rate))
      )
    }
  }
}
